File: clients/python/cloudmeter/_cost_projector.py
"""
Native Python cost projector — port of sidecar-rs/src/projector.rs.

CPU proxy: Python has no per-request CPU measurement API (GIL).
Wall-clock duration_ms / 1000 is used as cpu_core_seconds.
Cost accuracy is ±40% vs the Java agent's ±20%.
"""

import statistics
from typing import Dict, List, Optional

from ._pricing_catalog import (
    GIB_IN_BYTES,
    HOURS_PER_MONTH,
    InstanceType,
    get_egress_rate_per_gib,
    get_instances,
)

SCALE_USERS = [100, 200, 500, 1_000, 2_000, 5_000, 10_000, 20_000, 50_000, 100_000, 500_000, 1_000_000]
RECORDING_WINDOW_S = 300.0  # 5-minute rolling window (matches Rust sidecar)


def project(metrics: List[Dict], opts: Optional[Dict] = None) -> Dict:
    """
    Project monthly cost per route.

    opts keys:
        provider                  str   'AWS' | 'GCP' | 'AZURE'  (default 'AWS')
        target_users              int   (default 1000)
        requests_per_user_per_second  float  (default 1.0)
        budget_usd                float  (default 0 = disabled)

    Returns dict with keys: projections, warmup_count
    """
    opts = opts or {}
    provider = opts.get("provider", "AWS")
    target_users = opts.get("target_users", 1000)
    rpu = opts.get("requests_per_user_per_second", 1.0)
    budget_usd = opts.get("budget_usd", 0.0)

    instances = get_instances(provider)
    egress_rate = get_egress_rate_per_gib(provider)

    live = [m for m in metrics if not m.get("warmup", False)]
    warmup_count = len(metrics) - len(live)

    if not live:
        return {"projections": [], "warmup_count": warmup_count}

    total_observed_rps = len(live) / RECORDING_WINDOW_S
    safe_rps = max(total_observed_rps, 1e-9)
    target_total_rps = target_users * rpu

    by_route: Dict[str, list] = {}
    for m in live:
        by_route.setdefault(m["route"], []).append(m)

    projections = [
        _project_route(route, entries, safe_rps, target_total_rps, target_users, rpu, budget_usd, instances, egress_rate)
        for route, entries in by_route.items()
    ]
    projections.sort(key=lambda p: p["projected_monthly_cost_usd"], reverse=True)
    return {"projections": projections, "warmup_count": warmup_count}


def compute_stats(metrics: List[Dict], projections: List[Dict], opts: Optional[Dict] = None) -> List[Dict]:
    """
    Compute per-request cost distribution (p50/p95/p99) per route.
    Used by the variance panel in the dashboard.
    """
    opts = opts or {}
    provider = opts.get("provider", "AWS")
    egress_rate = get_egress_rate_per_gib(provider)
    instances = get_instances(provider)

    hourly_by_route: Dict[str, float] = {}
    for p in projections:
        inst = next((i for i in instances if i.name == p["recommended_instance"]), instances[0])
        hourly_by_route[p["route"]] = inst.hourly_usd

    live = [m for m in metrics if not m.get("warmup", False)]

    by_route: Dict[str, list] = {}
    for m in live:
        by_route.setdefault(m["route"], []).append(m)

    stats = []
    for route, entries in by_route.items():
        hourly_usd = hourly_by_route.get(route, 0.096)  # fallback: m5.large

        costs = sorted(
            (m["duration_ms"] / 1000.0 / 3600.0) * hourly_usd
            + (m.get("egress_bytes", 0) / GIB_IN_BYTES) * egress_rate
            for m in entries
        )

        p50 = _percentile(costs, 50)
        p95 = _percentile(costs, 95)
        p99 = _percentile(costs, 99)
        ratio = p95 / p50 if p50 > 0 else 0.0

        stats.append(
            {
                "route": route,
                "requests": len(entries),
                "p50_cost_usd": p50,
                "p95_cost_usd": p95,
                "p99_cost_usd": p99,
                "variance_ratio": round(ratio, 2),
                "variance_warning": ratio > 1.5,
            }
        )

    stats.sort(key=lambda s: s["p95_cost_usd"], reverse=True)
    return stats


# ── internal ──────────────────────────────────────────────────────────────────

def _project_route(
    route: str,
    entries: list,
    total_observed_rps: float,
    target_total_rps: float,
    target_users: int,
    rpu: float,
    budget_usd: float,
    instances: List[InstanceType],
    egress_rate: float,
) -> Dict:
    observed_rps = len(entries) / RECORDING_WINDOW_S
    scale_factor = target_total_rps / total_observed_rps
    projected_rps = observed_rps * scale_factor

    cpu_core_secs = [m["duration_ms"] / 1000.0 for m in entries]
    egress_bytes = [float(m.get("egress_bytes", 0)) for m in entries]
    durations = [float(m["duration_ms"]) for m in entries]

    median_cpu = _median(cpu_core_secs)
    median_egress = _median(egress_bytes)
    median_duration = _median(durations)

    monthly_cost = _monthly_cost(projected_rps, median_cpu, median_egress, instances, egress_rate)
    cost_per_user = monthly_cost / max(target_users, 1)
    recommended = _select_instance(projected_rps * median_cpu, instances)

    cost_curve = []
    for users in SCALE_USERS:
        scaled_rps = observed_rps * ((users * rpu) / total_observed_rps)
        cost_curve.append(
            {
                "users": users,
                "monthly_cost_usd": _r2(_monthly_cost(scaled_rps, median_cpu, median_egress, instances, egress_rate)),
            }
        )

    return {
        "route": route,
        "observed_rps": _r4(observed_rps),
        "projected_rps": _r1(projected_rps),
        "projected_monthly_cost_usd": _r2(monthly_cost),
        "projected_cost_per_user_usd": _r6(cost_per_user),
        "recommended_instance": recommended.name,
        "median_duration_ms": _r2(median_duration),
        "median_cpu_ms": _r2(median_cpu * 1000.0),
        "exceeds_budget": budget_usd > 0 and monthly_cost > budget_usd,
        "cost_curve": cost_curve,
    }


def _monthly_cost(
    projected_rps: float,
    median_cpu_core_sec: float,
    median_egress_bytes: float,
    instances: List[InstanceType],
    egress_rate: float,
) -> float:
    required_cores = projected_rps * median_cpu_core_sec
    inst = _select_instance(required_cores, instances)
    seconds_per_mo = HOURS_PER_MONTH * 3600.0
    egress_gib_per_mo = projected_rps * median_egress_bytes * seconds_per_mo / GIB_IN_BYTES
    return inst.hourly_usd * HOURS_PER_MONTH + egress_gib_per_mo * egress_rate


def _select_instance(required_cores: float, instances: List[InstanceType]) -> InstanceType:
    for inst in instances:
        if inst.vcpu >= required_cores:
            return inst
    return instances[-1]


def _median(values: List[float]) -> float:
    if not values:
        return 0.0
    return statistics.median(values)


def _percentile(sorted_values: List[float], p: int) -> float:
    if not sorted_values:
        return 0.0
    idx = min(int(p / 100.0 * len(sorted_values)), len(sorted_values) - 1)
    return sorted_values[idx]


def _r1(v: float) -> float:
    return round(v, 1)


def _r2(v: float) -> float:
    return round(v, 2)


def _r4(v: float) -> float:
    return round(v, 4)


def _r6(v: float) -> float:
    return round(v, 6)

File: clients/node/src/cost-projector.js
'use strict'

// Native Node.js cost projector — port of sidecar-rs/src/projector.rs.
//
// Key adaptation: the Java agent measures cpu_core_seconds via ThreadMXBean.
// Node.js has no equivalent, so durationMs is used as a proxy:
//   cpuCoreSec = durationMs / 1000
// This is a reasonable approximation for synchronous handlers.

const {
  PRICING_DATE,
  HOURS_PER_MONTH,
  GIB_IN_BYTES,
  getInstances,
  getEgressRatePerGib,
} = require('./pricing-catalog')

const SCALE_USERS        = [100, 200, 500, 1000, 2000, 5000, 10000, 20000, 50000, 100000, 500000, 1000000]
const RECORDING_WINDOW_S = 300  // 5-minute rolling window (matches Rust sidecar)

// ── project ───────────────────────────────────────────────────────────────────
//
// metrics: array of objects from reporter.getMetrics()
//   { route, method, status, durationMs, egressBytes, ts }
//
// opts:
//   provider                  'AWS' | 'GCP' | 'AZURE'   (default: 'AWS')
//   region                    string                      (default: 'us-east-1')
//   targetUsers               number                      (default: 1000)
//   requestsPerUserPerSecond  number                      (default: 1.0)
//   budgetUsd                 number                      (default: 0 = disabled)
//
// Returns: { projections, warmupCount }
//   projections sorted descending by projectedMonthlyCostUsd
//
function project(metrics, opts = {}) {
  const {
    provider                 = 'AWS',
    targetUsers              = 1000,
    requestsPerUserPerSecond = 1.0,
    budgetUsd                = 0,
  } = opts

  const instances  = getInstances(provider)
  const egressRate = getEgressRatePerGib(provider)

  // Partition warmup vs live using the flag set by reporter.js at record time
  const live        = metrics.filter(m => !m.warmup)
  const warmupCount = metrics.length - live.length

  if (live.length === 0) return { projections: [], warmupCount }

  const totalObservedRps = live.length / RECORDING_WINDOW_S
  const safeRps          = totalObservedRps > 0 ? totalObservedRps : 1.0
  const targetTotalRps   = targetUsers * requestsPerUserPerSecond

  // Group by route
  const byRoute = new Map()
  for (const m of live) {
    if (!byRoute.has(m.route)) byRoute.set(m.route, [])
    byRoute.get(m.route).push(m)
  }

  const projections = []
  for (const [route, entries] of byRoute) {
    projections.push(
      _projectRoute(route, entries, safeRps, targetTotalRps, targetUsers,
        requestsPerUserPerSecond, budgetUsd, instances, egressRate)
    )
  }

  projections.sort((a, b) => b.projectedMonthlyCostUsd - a.projectedMonthlyCostUsd)
  return { projections, warmupCount }
}

// ── computeStats ──────────────────────────────────────────────────────────────
//
// Computes per-request cost distribution (p50/p95/p99) per route.
// Used by the variance panel in the dashboard.
//
function computeStats(metrics, projections, opts = {}) {
  const { provider = 'AWS' } = opts
  const egressRate = getEgressRatePerGib(provider)
  const instances  = getInstances(provider)

  // Map route → hourlyUsd from the recommended instance in projections
  const hourlyByRoute = new Map()
  for (const p of projections) {
    const inst = instances.find(i => i.name === p.recommendedInstance) || instances[0]
    hourlyByRoute.set(p.route, inst.hourlyUsd)
  }

  const live = metrics.filter(m => !m.warmup)

  const byRoute = new Map()
  for (const m of live) {
    if (!byRoute.has(m.route)) byRoute.set(m.route, [])
    byRoute.get(m.route).push(m)
  }

  const stats = []
  for (const [route, entries] of byRoute) {
    const hourlyUsd = hourlyByRoute.get(route) || 0.096  // fallback: m5.large

    // Per-request cost = cpu cost + egress cost
    const costs = entries
      .map(m => {
        const cpuCost    = (m.durationMs / 1000 / 3600) * hourlyUsd
        const egressCost = ((m.egressBytes || 0) / GIB_IN_BYTES) * egressRate
        return cpuCost + egressCost
      })
      .sort((a, b) => a - b)

    const p50   = _percentile(costs, 50)
    const p95   = _percentile(costs, 95)
    const p99   = _percentile(costs, 99)
    const ratio = p50 > 0 ? p95 / p50 : 0

    stats.push({
      route,
      requests:       entries.length,
      p50CostUsd:     p50,
      p95CostUsd:     p95,
      p99CostUsd:     p99,
      varianceRatio:  _round2(ratio),
      varianceWarning: ratio > 1.5,
    })
  }

  stats.sort((a, b) => b.p95CostUsd - a.p95CostUsd)
  return stats
}

// ── internal ──────────────────────────────────────────────────────────────────

function _projectRoute(route, entries, totalObservedRps, targetTotalRps, targetUsers,
  rpu, budgetUsd, instances, egressRate) {

  const observedRps  = entries.length / RECORDING_WINDOW_S
  const scaleFactor  = targetTotalRps / totalObservedRps
  const projectedRps = observedRps * scaleFactor

  // durationMs / 1000 = cpu core-seconds proxy
  const cpuCoreSecs = entries.map(m => m.durationMs / 1000)
  const egressBytes = entries.map(m => m.egressBytes || 0)
  const durations   = entries.map(m => m.durationMs)

  const medianCpu      = _median(cpuCoreSecs)
  const medianEgress   = _median(egressBytes)
  const medianDuration = _median(durations)

  const monthlyCost = _monthlyCost(projectedRps, medianCpu, medianEgress, instances, egressRate)
  const costPerUser = monthlyCost / (targetUsers || 1)
  const recommended = _selectInstance(projectedRps * medianCpu, instances)

  const costCurve = SCALE_USERS.map(users => {
    const scaledTargetRps = users * rpu
    const scaledRps       = observedRps * (scaledTargetRps / totalObservedRps)
    return {
      users,
      monthlyCostUsd: _round2(_monthlyCost(scaledRps, medianCpu, medianEgress, instances, egressRate)),
    }
  })

  return {
    route,
    observedRps:             _round4(observedRps),
    projectedRps:            _round1(projectedRps),
    projectedMonthlyCostUsd: _round2(monthlyCost),
    projectedCostPerUserUsd: _round6(costPerUser),
    recommendedInstance:     recommended.name,
    medianDurationMs:        _round2(medianDuration),
    medianCpuMs:             _round2(medianCpu * 1000),
    exceedsBudget:           budgetUsd > 0 && monthlyCost > budgetUsd,
    costCurve,
  }
}

function _monthlyCost(projectedRps, medianCpuCoreSec, medianEgressBytes, instances, egressRate) {
  const requiredCores  = projectedRps * medianCpuCoreSec
  const inst           = _selectInstance(requiredCores, instances)
  const secondsPerMo   = HOURS_PER_MONTH * 3600
  const egressGibPerMo = projectedRps * medianEgressBytes * secondsPerMo / GIB_IN_BYTES
  return inst.hourlyUsd * HOURS_PER_MONTH + egressGibPerMo * egressRate
}

function _selectInstance(requiredCores, instances) {
  return instances.find(i => i.vCpu >= requiredCores) || instances[instances.length - 1]
}

function _median(arr) {
  if (!arr.length) return 0
  const s = arr.slice().sort((a, b) => a - b)
  const m = Math.floor(s.length / 2)
  return s.length % 2 === 0 ? (s[m - 1] + s[m]) / 2 : s[m]
}

function _percentile(sorted, p) {
  if (!sorted.length) return 0
  const idx = Math.min(Math.floor((p / 100) * sorted.length), sorted.length - 1)
  return sorted[idx]
}

function _round1(v) { return Math.round(v * 10) / 10 }
function _round2(v) { return Math.round(v * 100) / 100 }
function _round4(v) { return Math.round(v * 10000) / 10000 }
function _round6(v) { return Math.round(v * 1_000_000) / 1_000_000 }

module.exports = { project, computeStats, PRICING_DATE, SCALE_USERS }
